src/datasources/asset/AssetDataSource.ts

```diff
@@ -0,0 +1,325 @@
+    let arrayOfEntityIds: string[];
+    if (query.entityType === EntityType.ASSET) {
+      arrayOfEntityIds = replaceVariables([assetIdentifier], this.templateSrv);
+    } else {
+      arrayOfEntityIds = replaceVariables([minionId], this.templateSrv);
+    }
+    // filter empty values
+    arrayOfEntityIds = arrayOfEntityIds.filter(Boolean);
+    const resultArray: Array<{
+      id: string,
+      datetimes: number[],
+      values: number[]
+    }> = [];
+    if (!arrayOfEntityIds.length) {
+      if (query.entityType === EntityType.ASSET) {
+        let conditions: string[] = [];
+        if (workspace_id) {
+          conditions.push(`workspace = "${workspace_id}"`);
+        }
+        if (minion_id) {
+          conditions.push(`${AssetFilterProperties.LocationMinionId} = "${minion_id}"`);
+        }
+        const assetFilter = conditions.join(' and ');
+        const assetsResponse = await this.queryAssets(assetFilter, 25);
+        arrayOfEntityIds = assetsResponse.map((asset) => {
+          return asset.id;
+        })
+      } else {
+        let filterString = '';
+        if (query.workspace) {
+          filterString += `workspace = "${query.workspace}"`;
+        }
+        let systemMetadata = await this.post<{ data: SystemMetadata[] }>(this.sysmgmtUrl + '/query-systems', {
+          filter: filterString,
+          projection: `new(${defaultProjection.join()})`,
+          orderBy: defaultOrderBy,
+          take: 10
+        });
+        arrayOfEntityIds = systemMetadata.data.map((system) => {
+          return system.id;
+        })
+      }
+    }
+    for (let index = 0; index < arrayOfEntityIds.length; index++) {
+      let data: AssetUtilizationHistory[] = [];
+      let requestCount = 0;
+      do {
+        try {
+          let requestBody: QueryAssetUtilizationHistoryRequest = {
+            "utilizationFilter": utilizationFilter,
+            "continuationToken": continuationToken,
+            "orderBy": AssetUtilizationOrderBy.START_TIMESTAMP,
+            "orderByDescending": false
+          }
+          if (query.entityType === EntityType.ASSET) {
+            requestBody.assetFilter = `${AssetFilterProperties.AssetIdentifier} = "${arrayOfEntityIds[index]}"`;
+          } else {
+            requestBody.assetFilter = `${AssetFilterProperties.LocationMinionId} = "${arrayOfEntityIds[index]}" and ${AssetFilterProperties.IsSystemController} = true`;
+          }
+          const response: AssetUtilizationHistoryResponse = await this.post<AssetUtilizationHistoryResponse>(this.baseUrl + '/query-asset-utilization-history', requestBody);
+          requestCount++;
+          continuationToken = response.continuationToken || '';
+          data = data.concat(response.assetUtilizations);
+        } catch (error) {
+          throw new Error(`An error occurred while retrieving utilization history: ${error}`);
+        }
+      } while (continuationToken && (maxDataPoints && requestCount < maxDataPoints));
```

Comment:
`options.maxDataPoints` defaults to the width of the panel, which is usually 100s of pixels. Are we ever actually making that many requests here? That seems weird to me.